### PR TITLE
[PLAT-8819] Add `leaveNetworkRequestBreadcrumbForTask:metrics:`

### PR DESCRIPTION
**Pull request by nickdowell**
## Goal

Facilitate leaving network request breadcrumbs without relying on the plugin's Objective-C method swizzling.
* https://github.com/bugsnag/bugsnag-cocoa/issues/1420

## Design

`+[Bugsnag leaveNetworkRequestBreadcrumbForTask:metrics:]` provides a public API for leaving a network request breadcrumb based on an `NSURLSessionTask` and `NSURLSessionTaskMetrics`.

## Changeset

Adds `+[Bugsnag leaveNetworkRequestBreadcrumbForTask:metrics:]` and `-[BugsnagClient leaveNetworkRequestBreadcrumbForTask:metrics:]`.

Moves breadcrumb message & metadata computation from `BSGURLSessionTracingDelegate` to `BSGNetworkBreadcrumb.m`, called from `-[BugsnagClient leaveNetworkRequestBreadcrumbForTask:metrics:]`.

Calls `-[BugsnagClient leaveNetworkRequestBreadcrumbForTask:metrics:]` from `-[BSGURLSessionTracingDelegate URLSession:task:didFinishCollectingMetrics:]`.

## Testing

Migrates and modifies existing unit tests to continue passing.

Verified by existing E2E scenario.

### CODE REVIEW

File: Bugsnag/Breadcrumbs/BSGNetworkBreadcrumb.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#import <Bugsnag/Bugsnag.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+API_AVAILABLE(macosx(10.12), ios(10.0), watchos(3.0), tvos(10.0))
+BugsnagBreadcrumb * _Nullable BSGNetworkBreadcrumbWithTaskMetrics(NSURLSessionTask *task, NSURLSessionTaskMetrics *metrics);
+
+NSDictionary<NSString *, id> * _Nullable BSGURLParamsForQueryItems(NSArray<NSURLQueryItem *> * _Nullable queryItems);
+
+NSString * _Nullable BSGURLStringForComponents(NSURLComponents * _Nullable URLComponents);
```
Comment (kstenerud) [inline Bugsnag/Breadcrumbs/BSGNetworkBreadcrumb.h:18]: Do these two methods need to be public?

File: Bugsnag/Breadcrumbs/BSGNetworkBreadcrumb.m [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    // NSURLSessionTaskTransactionMetrics.response is nil when a custom NSURLProtocol is used. 
+    // If there was an error, task.response will be nil.
+    NSURLResponse *response = task.response; 
+    if ([response isKindOfClass:[NSHTTPURLResponse class]]) {
+        NSInteger statusCode = ((NSHTTPURLResponse *)response).statusCode;
+        if (100 <= statusCode && statusCode < 400) {
+            message = @"NSURLSession request succeeded";
+        }
+        if (400 <= statusCode && statusCode < 500) {
+            message = @"NSURLSession request failed";
```
Comment (kstenerud) [inline Bugsnag/Breadcrumbs/BSGNetworkBreadcrumb.m:46]: Do we put any message when we get an error response (500)?

File: Bugsnag/Breadcrumbs/BSGNetworkBreadcrumb.h (hunk 1 above)
Comment (nickdowell) [reply Bugsnag/Breadcrumbs/BSGNetworkBreadcrumb.h:18]: They need to be declared in this non-public header for unit testing.

File: Bugsnag/Breadcrumbs/BSGNetworkBreadcrumb.m (hunk 2 above)
Comment (nickdowell) [reply Bugsnag/Breadcrumbs/BSGNetworkBreadcrumb.m:46]: Yes - `message` is initialized with "NSURLSession request error" on line 35.

**Review by kstenerud — APPROVED** [review]

**Comment by eseay** [comment]
Thank you for doing this! A great addition!

**Pull request merged**